layer/layer_data.h

```diff
@@ -62,67 +144,91 @@ class LayerData {
     }
   }
 
-  // Records the dispatch table that is associated with |instance|.
+  // Records the dispatch table and instance key that is associated with
+  // |instance|.
   bool AddInstance(VkInstance instance,
                    const VkLayerInstanceDispatchTable& dispatch_table) {
+    InstanceKey key(instance);
+
     absl::MutexLock lock(&instance_dispatch_lock_);
-    return instance_dispatch_map_.insert({instance, dispatch_table}).second;
+    const bool inserted =
+        instance_dispatch_map_.insert({key, dispatch_table}).second;
+    if (inserted) {
+      instance_keys_map_[key] = instance;
+    }
+    return inserted;
   }
 
   // Removes the dispatch table and physical devices associated with |instance|.
   void RemoveInstance(VkInstance instance);
 
-  // Records that the physical device |gpu| is associated with |instance|.
-  void AddPhysicalDevice(VkInstance instance, VkPhysicalDevice gpu) {
-    absl::MutexLock lock(&gpu_instance_lock_);
-    assert((gpu_instance_map_.count(gpu) == 0 ||
-            gpu_instance_map_[gpu] == instance) &&
-           "PhysicalDevice already bound to another Instance");
-    gpu_instance_map_[gpu] = instance;
-  }
-
-  // Returns the instance associated with |gpu|, or a null handle if there is
-  // none.
-  VkInstance GetInstance(VkPhysicalDevice gpu) const {
-    absl::MutexLock lock(&gpu_instance_lock_);
-    if (auto it = gpu_instance_map_.find(gpu); it != gpu_instance_map_.end())
+  // Returns the instance associated with |instance_key|, or a null handle if
+  // there is none.
+  VkInstance GetInstance(InstanceKey instance_key) const {
```

Comment:
Since `AddInstance` and `RemoveInstance` take a `VkInstance`, can we have `GetInstance` also take a `VkInstance` and constructs the `InstanceKey` inside?

Reply:
`GetInstance` is supposed to map from instance keys to VKInstances, passing a VkInstance will get you the same object back. I think it would make more sense to overload it to accept VkPhysicalDevice instead, but that circumvents the InstanceKey abstraction.
What do you think?

Reply:
Ah sorry you're right, it doesn't make sense to pass VkInstance. It's fine to leave it as it is.